stem/descriptor/hidden_service.py

```diff
@@ -0,0 +1,1350 @@
+  :var str onion_key: public key for communicating with this introduction point
+  :var str service_key: public key for communicating with this hidden service
+  :var list intro_authentication: tuples of the form (auth_type, auth_data) for
+    establishing a connection
+  """
+
+
+class IntroductionPointV3(collections.namedtuple('IntroductionPointV3', ['link_specifiers', 'onion_key_raw', 'auth_key_cert', 'enc_key_raw', 'enc_key_cert', 'legacy_key_raw', 'legacy_key_cert'])):
+  """
+  Introduction point for a v3 hidden service.
+
+  .. versionadded:: 1.8.0
+
+  :var list link_specifiers: :class:`~stem.client.datatype.LinkSpecifier` where this service is reachable
+  :var str onion_key_raw: base64 ntor introduction point public key
+  :var stem.descriptor.certificate.Ed25519Certificate auth_key_cert: cross-certifier of the signing key with the auth key
+  :var str enc_key_raw: base64 introduction request encryption key
+  :var stem.descriptor.certificate.Ed25519Certificate enc_key_cert: cross-certifier of the signing key by the encryption key
+  :var str legacy_key_raw: base64 legacy introduction point RSA public key
+  :var str legacy_key_cert: base64 cross-certifier of the signing key by the legacy key
+  """
+
+  @staticmethod
+  def parse(content):
+    """
+    Parses an introduction point from its descriptor content.
+
+    :param str content: descriptor content to parse
+
+    :returns: :class:`~stem.descriptor.hidden_service.IntroductionPointV3` for the descriptor content
+
+    :raises: **ValueError** if descriptor content is malformed
+    """
+
+    entry = _descriptor_components(content, False)
+    link_specifiers = IntroductionPointV3._parse_link_specifiers(_value('introduction-point', entry))
+
+    onion_key_line = _value('onion-key', entry)
+    onion_key = onion_key_line[5:] if onion_key_line.startswith('ntor ') else None
+
+    _, block_type, auth_key_cert = entry['auth-key'][0]
+    auth_key_cert = Ed25519Certificate.from_base64(auth_key_cert)
+
+    if block_type != 'ED25519 CERT':
+      raise ValueError('Expected auth-key to have an ed25519 certificate, but was %s' % block_type)
+
+    enc_key_line = _value('enc-key', entry)
+    enc_key = enc_key_line[5:] if enc_key_line.startswith('ntor ') else None
+
+    _, block_type, enc_key_cert = entry['enc-key-cert'][0]
+    enc_key_cert = Ed25519Certificate.from_base64(enc_key_cert)
+
+    if block_type != 'ED25519 CERT':
+      raise ValueError('Expected enc-key-cert to have an ed25519 certificate, but was %s' % block_type)
+
+    legacy_key = entry['legacy-key'][0][2] if 'legacy-key' in entry else None
+    legacy_key_cert = entry['legacy-key-cert'][0][2] if 'legacy-key-cert' in entry else None
+
+    return IntroductionPointV3(link_specifiers, onion_key, auth_key_cert, enc_key, enc_key_cert, legacy_key, legacy_key_cert)
+
+  @staticmethod
+  def create(address, port, expiration = None, onion_key = None, enc_key = None, auth_key = None, signing_key = None):
```

Comment:
Hmm, seems like the constructor here only allows us to specify address/port, but usually HSv3 descriptor introduction points contain other sorts of link specifiers too (particularly fingeprrint and ed25519). Would it be possible to support more kinds of link specifiers here?

Reply:
Yup! I wasn't sure what API would be most useful for callers so I put this together as a demo. As the pydoc says...

"Simplified constructor. For more sophisticated use cases you can use this as a template for how introduction points are properly created."

What API would you like?

Reply:
Hmm, I'm not quite sure. Perhaps we can pass a list of link specifiers directly to the simplfied constructor? Or allow more arguments like `ed25519_identity` and `fingerprint` to the simplified constructor? Or I can use the non-simplified constructor if this is possible?

I think I'm OK with all the above as long as the functionality is possible in the end. What would you prefer?